pkg/v1/layout/write.go

```diff
@@ -221,43 +221,83 @@ func (l Path) WriteFile(name string, data []byte, perm os.FileMode) error {
 // blobs/{hash.Algorithm}/{hash.Hex}.
 func (l Path) WriteBlob(hash v1.Hash, r io.ReadCloser) error {
+	return l.writeBlob(hash, 0, r, nil)
+}
+
+func (l Path) writeBlob(hash v1.Hash, size int64, r io.ReadCloser, renamer func() (v1.Hash, error)) error {
 	dir := l.path("blobs", hash.Algorithm)
 	if err := os.MkdirAll(dir, os.ModePerm); err != nil && !os.IsExist(err) {
 		return err
 	}
 
+	// Check if blob already exists and is the correct size
 	file := filepath.Join(dir, hash.Hex)
-	if _, err := os.Stat(file); err == nil {
-		// Blob already exists, that's fine.
+	if s, err := os.Stat(file); err == nil && (s.Size() == size || size == 0) {
 		return nil
 	}
-	w, err := os.Create(file)
+
+	// If a renamer func was provided write to a temporary file
+	open := func() (*os.File, error) { return os.Create(file) }
+	if renamer != nil {
+		open = func() (*os.File, error) { return ioutil.TempFile(dir, hash.Hex) }
+	}
+	w, err := open()
 	if err != nil {
 		return err
 	}
 	defer w.Close()
 
-	_, err = io.Copy(w, r)
-	return err
-}
+	// Write to file and optionally rename
+	if _, err := io.Copy(w, r); err != nil || renamer == nil {
+		return err
+	}
+	finalHash, err := renamer()
+	if err != nil {
+		return err
+	}
+	renamePath := l.path("blobs", finalHash.Algorithm, finalHash.Hex)
 
-// TODO: A streaming version of WriteBlob so we don't have to know the hash
-// before we write it.
+	// Always close file before renaming
+	if err := w.Close(); err != nil {
+		return err
+	}
+	return os.Rename(w.Name(), renamePath)
+}
 
-// TODO: For streaming layers we should write to a tmp file then Rename to the
-// final digest.
+// writeLayer writes the compressed layer to a blob. Unlike WriteBlob it will
+// write to a temporary file (suffixed with .tmp) within the layout until the
+// compressed reader is fully consumed and written to disk. Also unlike
+// WriteBlob, it will not skip writing and exit without error when a blob file
+// exists, but does not have the correct size. (The blob hash is not
+// considered, because it may be expensive to compute.)
 func (l Path) writeLayer(layer v1.Layer) error {
 	d, err := layer.Digest()
 	if err != nil {
-		return err
+		// Allow digest errors, since streams may not have calculated the hash
+		// yet. Instead, use an empty value, which will be transformed into a
+		// random file name with `ioutil.TempFile` and the final digest will be
+		// calculated after writing to a temp file and before renaming to the
+		// final path.
+		d = v1.Hash{Algorithm: "sha256", Hex: ""}
+	}
+
+	s, err := layer.Size()
+	if err != nil {
+		// Allow size errors, since streams may not have calculated the size
+		// yet. Instead, use zero as a sentinel value meaning that no size
+		// comparison can be done and any sized blob file should be considered
+		// valid and not overwritten.
+		//
+		// TODO: Provide an option to always overwrite blobs.
+		s = 0
```

Comment:
We've used `-1` as a sentinel value in similar situations: https://github.com/google/go-containerregistry/blob/7adcadef6bc8adf07cc587a5a5a443afe7603d54/internal/verify/verify.go#L32

A zero-sized blob seems maybe reasonable, but a `-1` sized blob doesn't make much sense, so I think it's a better sentinel value.